### DISCUSSION TITLE
tesr to mesh

### DISCUSSION THREAD

**Discussion by ThomasChauve** [neperfepx/neper#207] [Q&A] [answered]
Hello,
Thanks for the tools you provide. I was trying to do transform a tesr to mesh and I got an error. 
My tesr is a custom one build for a VTK file. Therefore it can be no written correctly.
Anyway I got this error :

```
neper -M CI02_539x495_micro.tesr
```

```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.1.2
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt
Info   : Running on 16 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2020, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/chauvet//.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] CI02_539x495_micro.tesr
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading raster tessellation...
Info   :     [i] Parsing file `CI02_539x495_micro.tesr'...
Info   :     [i] Parsed file `CI02_539x495_micro.tesr'.
Info   :   - Reconstructing topology...
Info   :     > Reconstructing 1D mesh... 100%      
Info   :     > Reconstructing 0D mesh... 100%      
Error  : You have discovered a bug in Neper!  Please file an issue at
         https://github.com/rquey/neper/issues.  Thank you.
```

Here is the input file. I just rename the '.tesr' ' to '.txt' as github was not ok to take 'tesr'
[CI02_539x495_micro.txt](https://github.com/rquey/neper/files/6425972/CI02_539x495_micro.txt)

Therefore I have a couple of question :

1. Is it possible to give a VTK as input in neper ?
2. Is it possible to build a mesh from a 3D raster file ? (from the doc it seems not possible)
3. Overwise do you have over tools to recommand to mesh a 2D/3D microstructure for an experimental raster image ?

Thanks for the great work.
Best
Thomas Chauve

PS: I am using Ubuntu 20.04 and Neper 4.1.2

**Comment by rquey** [comment, marked as answer]
You just have to filter your tesr before meshing to solve topological issues, using `-transform rmsat,grow`:
```
$ neper -T -loadtesr CI02_539x495_micro.txt -transform "rmsat,grow" -o filtered
$ neper -M filtered.tesr -rcl 0.5 -mesh2dalgo fron,dela
$ neper -V filtered.tesr,filtered.msh -datavoxedgerad 0 -datacellcol id -dataelsetcol id -imagesize 600:600 -cameraangle 12 -showelt1d all -dataelt1drad 0.2 -showtesr 1 -showmesh 0 -print filtered-tesr -showtesr 0 -showmesh 1 -print filtered-mesh
```
(`-transform` changes only 3 pixels...)

![image](https://user-images.githubusercontent.com/2659376/117143074-b5bf6200-adb0-11eb-816f-6fd38f861566.png)
![image](https://user-images.githubusercontent.com/2659376/117143095-b9eb7f80-adb0-11eb-9032-f2e077e7296c.png)

Neper does not take VTK files as input, and it can mesh a 3D tesr by approximating it as a tess (`-morpho tesr:file()`), see [Quey and Renversade, 2018](https://neper.info/downloads.html#references).

**Comment by ThomasChauve** [comment]
Hello,

Thanks for your quick answer. It works fine.

I just have one more question about the transforming tesr to tess in a 3D case ?
How do you do that ?

I tried from an home made tesr :
[CI02_539x495_micro.txt](https://github.com/rquey/neper/files/6433433/CI02_539x495_micro.txt)

First to filter it
`neper -T -loadtesr CI02_539x495_micro.txt -transform "rmsat,grow" -o filtered_CI02`

Then to transformed it to 'tess' using something like :
`[neper -T -loadtesr filtered_CI02.tesr -morpho filtered_CI02.tesr -format tess`

But it doesn't work. It give me an empty file. I think I am not using correctly the -morpho option.
[filtered_CI02.txt](https://github.com/rquey/neper/files/6433486/filtered_CI02.txt)

Here is a visualisation of the filtered `tesr` file :
`neper -V filtered_CI02.tesr  -datacellcol id -showmesh 0 -print test`

![test](https://user-images.githubusercontent.com/9338747/117278233-e1018a00-ae60-11eb-9749-597df22dbb91.png)

Thanks for your help.
Thomas

**Reply by rquey** [reply]
For a general 3D polycrystal, use `-n from_morpho -morpho "tesr:file(name_of_tesr_file)" -domain "cube(...,...,...)"`:
- `-morpho` defines the morphology as the tesr.
- `-n from_morpho` defines the number of cells from the tesr.
- `-domain cube` defines the domain, which must be the same as the one of the tesr.

The tess (and mesh) will not look exactly like your tesr, because there is a convex-cell approximation.

In the case you showed, the tesr is very thin along z. It would be best to mesh it as I said before and extrude the mesh along z.

PS. You do not really need to filter before using `-morpho "tesr"`.

**Reply by janzenchoi** [reply]
Hello rquey,

Does this approach work for volumes with small width cells? I generated a `.tesr` file to simulate a volume with twin structures, and was unable to maintain the twin structures when converting to a `.tess` file. I have attached a photo of their respective visualisations.

![image](https://user-images.githubusercontent.com/75835751/176716446-f0d9667e-f304-48b6-a483-9cfb98b808fb.png)

Can you suggest an alternate approach? Thank you in advance.

Regards,
Janzen

**Reply by mkasemer** [reply]
@jazzzmannn : May I suggest simply making a tessellation, and integrating lamella within each grain in the tessellation in an attempt to represent annealing twins?

I think the issue is that Neper is going to view the regions in the tesr as separate grains, rather than lamella.

**Reply by janzenchoi** [reply]
Hello @mkasemer,

Thank you very much for your response.

I tried going down that road, but was unsuccessful in defining the lamellas based on a distribution of their widths and orientations. Is there a way to do so?

On that note, do you know the format for defining the twin structures as if they were lamella (i.e., non-separate grains) in the `.tesr` file? My script, like you mentioned, generates the `.tesr` file by treating the twin structures like separate grains.

**Reply by mkasemer** [reply]
@jazzzmannn : Regarding your second point, I do not know if Neper will consider multilevel tessellations in tesr for input. @rquey would have to answer that.

Regarding your first point, I likewise do not know if Neper can handle this general problem natively (currently). However, you can specify the lamellar width / orientation per-grain via external files. See Figure 5 in https://doi.org/10.1016/j.jmps.2017.03.013, for example (note: Neper commands have changed tremendously since those tessellations were created, so I cannot be of help with that example, this is purely illustrative). This application was for a second phase in Ti64, but can also be considered as annealing twins.

My suggestion is to create a tessellation / orientations (1 - level), calculate externally the lamellar widths / orientations for each grain, in whatever way you desire, and create the necessary normals / widths files for Neper. Then, create a multilevel tessellation with the same 1st level definition, with the second level defined as lamella with attributes loaded from file.

**Reply by janzenchoi** [reply]
Interesting. I will try externally calculating the lamellar widths and orientations. Thanks again for the quick and insightful responses, @mkasemer 👍. I really appreciate them.

**Reply by janzenchoi** [reply]
Hello @mkasemer,

Just one more question. Do you think Neper is capable of simulating lamellas of shape similar to the twin structures shown below?

![image](https://user-images.githubusercontent.com/75835751/176813544-d248c060-cff1-4342-b4dd-a1a84529964c.png)

If not, do you have any suggestions?

Thanks again,
Janzen

**Reply by rquey** [reply]
> Hello rquey,
> 
> Does this approach work for volumes with small width cells? I generated a `.tesr` file to simulate a volume with twin structures, and was unable to maintain the twin structures when converting to a `.tess` file. I have attached a photo of their respective visualisations.
> 
> ![image](https://user-images.githubusercontent.com/75835751/176716446-f0d9667e-f304-48b6-a483-9cfb98b808fb.png)
> 
> Can you suggest an alternate approach? Thank you in advance.
> 
> Regards, Janzen

From Neper's perspective, this is a very interesting case. How did you get these data?

The issue is not about the "small width cells", but about the fact that this is a multiscale polycrystal. Formally, this not a *normal* tessellation, and therefore a single-scale tessellation (which is *normal*) won't be able to reproduce it very well. See the [2018 CMAME paper](https://neper.info/papers.html#id2).

As @mkasemer said, you need multiscale for this, but, first, I'm curious about how you obtained the data.

**Reply by janzenchoi** [reply]
Hey @rquey,

Thanks for the reference.

I developed the volume (with the twins) in Dream3D and exported the voxel coordinates. I then wrote a script to create the .tesr file based on the voxel coordinates.

**Reply by rquey** [reply]
Okay, so, it's purely simulation? What type of polycrystal is it? Grain growth then "geometrical" twins?

**Reply by janzenchoi** [reply]
I'm pretty new to this, so I am not too sure. The volume is meant to be generated by specifying various statistics. The grains are generated first, then the twins, with each having their own set of statistics to specify.

I am more than happy to share the specifics of my process if it interests you.

**Reply by mkasemer** [reply]
@rquey: yes, it sounds idealized. I think what Dream3D does is essentially what I suggested: 

- builds a 1-level tessellation (grain growth, or whatever), giving a set of grains with corresponding crystallographic orientations,
- then for each grain calculates the twin normal and orientation based on the grain's orientation,
- finally introduces a 2-level tessellation into the first with tightly coupled lamella morphology coupled to the 1-level grain state.

I think this could be automated (certainly using external shell/python scripts), but it is system-dependent - i.e., grains in which the twins appear, the twins that appear (normal and xtal ori), etc. will all depend on the crystal symmetry and the material behavior. Here, I don't know what Dream3D is doing: @jazzzmannn, what aspects do you have to feed it for it to generate this microstructure? At minimum, does it know crystal type / material system?

**EDIT:** I should say that I find annealing twins deceptively complex. There are many twin variations that can form, and which ones that do form are going to be tightly coupled to the crystal symmetry, geometric state of the parent grains, crystallographic texture, material behavior, and the thermomechanical processing route. That is to say, it seems difficult to make a general rule that says "grain X should definitively have twin Y" (and I'm curious how exactly Dream3D makes this call). I would suggest writing a script that does its best to mimic the character of a witnessed microstructure, in terms of the coupling between parents/twins and the statistics concerning which twin variants are more likely (i.e., if there is a bias in terms of a certain variant(s)).

**Reply by rquey** [reply]
Thanks @mkasemer.

@jazzzmannn  - This is not very different from what Neper does, so, if the next step is to approximate the tesr by a Neper tess, I'd simply suggest to work with Neper from the beginning, and directly generate the tess (using external files for what Neper doesn't do natively).

**Reply by janzenchoi** [reply]
Hello @mkasemer and @rquey,

Sorry for the late reply. The input for generating the microstructure includes various distributions for the grain sizes, grain sphericities, crystallographic orientations, number of twins (per grain), twin thickness, and the twin fraction (i.e., how much of the grain does the twin section).

I think the best route for me now is what you both suggested, in that I should just use Neper for the entire process. Thank you for answering my questions and for the recommendation.

By the way, is Neper able to generate lamellas that don't fully section the grain?

**Reply by rquey** [reply]
Yes, apply several widths. Example:
```
neper -T -n 3::from_morpho -morpho "voronoi::lamellar(w=0.5:0.05:1,pos=start)" -o d
neper -V d.tess -datacellcol "scaleid(1)" -datacelltrs 0.5 -print img
```
![image](https://user-images.githubusercontent.com/2659376/177178538-42b9d473-cfab-42f1-990e-3c72dc0dc107.png)

It will take some tricks (using external files) to do what you want.
